openwisp_utils/admin_theme/context_processor.py

```diff
@@ -21,15 +20,22 @@ def build_menu(request=None):
     custom_items = getattr(settings, 'OPENWISP_ADMIN_MENU_ITEMS', [])
     items = custom_items or default_items
     menu = []
+    permitted_models = set()
+    if request and request.user.is_authenticated:
+        if hasattr(request.user, 'get_permissions'):
+            permissions = request.user.get_permissions
+        else:
+            permissions = request.user.get_all_permissions()
+        for permission in permissions:
+            permitted_models.add(permission.split('_')[-1])
     # loop over each item to build the menu
     # and check user has permission to see each item
     for item in items:
         app_label, model = item['model'].split('.')
         model_class = registry.apps.get_model(app_label, model)
         url = reverse('admin:{}_{}_changelist'.format(app_label, model.lower()))
         label = item.get('label', model_class._meta.verbose_name_plural)
-        model_admin = site._registry[model_class]
-        if not request or model_admin.has_module_permission(request):
+        if not request or (permitted_models and model.lower() in permitted_models):
```

Comment:
this is fragile, there could be two models using the same name but in different apps.

This would be better and more elegant:

```
request.user.has_perm(item['model'])
```

But it generates one query for every request, I think we can override `has_perm` in openwisp-users to make it use our cached permissions.